digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java

```diff
@@ -288,6 +287,38 @@ public boolean isAnyNotDoneAttempts()
     }
 
     @Override
+    public List<StoredSessionAttempt> findAttemptsCreatedBeforeWithState(Instant createdBefore, int state, long lastId, int limit)
+    {
+        return autoCommit((handle, dao) -> dao.findAttemptsCreatedBeforeWithState(sqlTimestampOf(createdBefore), state, lastId, limit));
+    }
+
+    @Override
+    public List<TaskAttemptSummary> findTasksStartedBeforeWithStateAndType(TaskType type, TaskStateCode[] states, Instant startedBefore, long lastId, int limit)
+    {
+        return autoCommit((handle, dao) ->
+                handle.createQuery(
+                    "select id, attempt_id, state" +
+                    " from tasks" +
+                    " where task_type = :type " +
```

Comment:
How about adding `(state_flags & 1) = 0` condition so that `WorkflowExecutionTimeoutEnforcer.enforceTaskTTLs` doesn't have to check same tasks again and again when an attempt includes a non-polling task which doesn't finish immediately?:

```
... +
// exclude already cancel-requested tasks
" and " + bitAnd("state_flags", Integer.toString(TaskStateFlags.CANCEL_REQUESTED)) + " = 0" +
```